bookshelf/src/controllers.php

```diff
@@ -0,0 +1,150 @@
+$app->post('/books/add', function (Request $request) use ($app) {
+    /** @var DataModelInterface $model */
+    $model = $app['bookshelf.model'];
+    /** @var ImageStorage $storage */
+    $storage = $app['bookshelf.storage'];
+    $files = $request->files;
+    $book = $request->request->all();
+    $image = $files->get('image');
+    if ($image && $image->isValid()) {
+        $book['imageUrl'] = $storage->storeFile(
+            $image->getRealPath(),
+            $image->getMimeType()
+        );
+    }
+    $id = $model->create($book);
+
+    return $app->redirect("/books/$id");
+});
+
+$app->get('/books/{id}', function ($id) use ($app) {
+    /** @var DataModelInterface $model */
+    $model = $app['bookshelf.model'];
+    $book = $model->read($id);
+    if (!$book) {
+        return new Response('', 404);
+    }
+    /** @var Twig_Environment $twig */
+    $twig = $app['twig'];
+
+    return $twig->render('view.html.twig', array('book' => $book));
+});
+
+$app->get('/books/{id}/edit', function ($id) use ($app) {
+    /** @var DataModelInterface $model */
+    $model = $app['bookshelf.model'];
+    $book = $model->read($id);
+    if (!$book) {
+        return new Response('', 404);
+    }
+    /** @var Twig_Environment $twig */
+    $twig = $app['twig'];
+
+    return $twig->render('form.html.twig', array(
+        'action' => 'Edit',
+        'book' => $book,
+    ));
+});
+
+$app->post('/books/{id}/edit', function (Request $request, $id) use ($app) {
+    $book = $request->request->all();
+    $book['id'] = $id;
+    /** @var ImageStorage $storage */
+    $storage = $app['bookshelf.storage'];
+    /** @var DataModelInterface $model */
+    $model = $app['bookshelf.model'];
+    $files = $request->files;
+    $image = $files->get('image');
+    // If the user uploaded a new image, we have to clean up the old image.
+    $oldImageUrl = '';
+    if ($image && $image->isValid()) {
+        $oldBook = $model->read($id);
+        if (isset($oldBook['imageUrl'])) {
+            $oldImageUrl = $oldBook['imageUrl'];
+        }
+        $book['imageUrl'] = $storage->storeFile(
+            $image->getRealPath(),
+            $image->getMimeType()
+        );
+    }
+    if ($model->update($book)) {
+        if ($oldImageUrl) {
+            $storage->deleteFile($oldImageUrl);
+        }
+
+        return $app->redirect("/books/$id");
+    }
+    return new Response('', 404);
+});
+
+$app->post('/books/{id}/delete',
+    function ($id) use ($app) {
+        /** @var DataModelInterface $model */
+        $model = $app['bookshelf.model'];
+        $book = $model->read($id);
+        if ($book) {
+            $model->delete($id);
+            /** @var ImageStorage $storage */
+            $storage = $app['bookshelf.storage'];
+            $imageUrl = isset($book['imageUrl']) ? $book['imageUrl'] : null;
+            if ($imageUrl) {
+                $storage->deleteFile($imageUrl);
+            }
+
+            return $app->redirect('/books/', 303);
```

Comment:
Is it possible to use the constants in Symfony's Response class?

https://github.com/symfony/symfony/blob/master/src/Symfony/Component/HttpFoundation/Response.php#L23